rows: Support all optional interfaces for driver.Rows

tcolgate opened the pull request:
This uses some code generation to wrap an incoming rows in a type
that supports the exact set of interfaces that are supported by
the original type.

tcolgate commented:
This adapts a technique used in prometheus client_golang but uses code generation due to the number of combinations of optional types permitted by driver.Rows.
Unfortunately this approach does seem to be needed to at least the mysql,pgx and sqlite3 drivers which all support different combinations of these optional interfaces.

tcolgate commented:
This poses a further interesting challenge. If the user wraps the Rows type (we are doing this with the goal of instrumenting row reads and open / close duration), The type we get back is itself wrapped and suffers from the same problem as this PR solves for the sqlmw rows type.
  One option would be to allow user wrapped Rows to implement an `func RowsUnwrap() driver.Rows`, that returns the original driver.Rows. We could check for that, call it if present and use the returned rows to probe the supported methods for wrapping.
    Users would then implement all the possible methods on their wrapper (passing calls through if present), but sqlmw would only expose a rows that supports the capabilities of the driver. 
  It might be better to leave that to a separate PR, to discuss API.

inconshreveable commented on tools/rows_picker_gen.go:
```diff
@@ -0,0 +1,149 @@
+}
+
+/*
+const (
+	rowsNextResultSet = 1 << iota
+	rowsColumnTypeDatabaseTypeName
+	rowsColumnTypeLength
+	rowsColumnTypeNullable
+	rowsColumnTypePrecisionScale
+	rowsColumnTypeScanType
+)
```
remove?

tcolgate replied:
> Thank you for this contribution! Can we add a few tests that exercise the new functionality and prove its correctness?

yes, I think so. The Go mysql, pgx and sqlite3 drivers expose different combiantions of these, so I was thinking of having mock drivers that match what they use. 
  It'll take me a while to get the tests sorted, didn't want to start the effort until I was sure you'd be interested in the merge.

inconshreveable commented on rows_picker.go:
```diff
@@ -0,0 +1,1018 @@
+package sqlmw
```
this file should have an appropriate generation comment at the top of it

inconshreveable replied:
sure, but the comment should indiciate that the file was generated, the command that generated it and not to edit the file manually

inconshreveable commented on rows_picker.go:
```diff
@@ -0,0 +1,1018 @@
+	}
+}
+
+type RowsUnwrapper interface {
+	RowsUnwrap() driver.Rows
+}
```
this can be defined in a non-generated file. it also needs to be documented for users of the library

inconshreveable reviewed (commented):
Thank you for this contribution! Can we add a few tests that exercise the new functionality and prove its correctness?

tcolgate commented:
Can you have a read of the RowsUnwrapper comment, I think it could be improved.
Assuming you are happy with these changes I'll start working on the tests tomorrow.

fho commented on rows.go:
```diff
@@ -3,19 +3,32 @@ package sqlmw
+// by the drivers that will be used with it. To support any arbitrary driver
+// all the optional methods must be supported.
+type RowsUnwrapper interface {
+	RowsUnwrap() driver.Rows
```
```suggestion
	Unwrap() driver.Rows
```

tcolgate replied:
Since this explicitly unwraps the rows type. and similar techniques could be applied elsewhere (like Tx) , being explicit felt like it made more sense here.

inconshreveable replied:
@fho's suggestion here feels right to me

re the comment: it could use some work, but it's directionally right.
- let's explain *why* you would wrap a Rows object
- "sqlmw will then pass a type that matches the optional interface set of the driver.Rows implementation of the original driver." this is hard to parse and could use clarity, i don't understand what it's trying to say
- maybe give an example method for the optional methods when it comes to wrapping rows objects

tcolgate commented:
Updated the doc string. Better I think, but if you have specific suggestions...

fho commented:
Thanks a lot for the PR. That is a great improvement.

1. According to the README Golang 1.9 is supported (which actually isn't in master: `./interceptor.go:17:36: undefined: driver.Connector`).
The rows_picker_gen.go from this PR is also not compatible with Go 1.9:
    ```go
    tools/rows_picker_gen.go:77:10: invalid operation: 1 << i (shift count type int, must be unsigned integer)
    tools/rows_picker_gen.go:85:12: invalid operation: 1 << len(intfs) (shift count type int, must be unsigned integer)
    ```
    This is trivial to fix.
Imho support for ancient go versions should be dropped.
Compatibility with the last 2-4 versions should be sufficient.
This is another topic though. :-)

2. I like the `Unwrap` approach.
If I understand the idea right, we could introduce in follow-up PRs similiar `TxUnwrapper`, `StmtUnwrapper`, `ConnUnwrapper`, `ResultUnwrapper` interfaces to make it work for all interceptor consistently? 
To make the same approach work for Stmts, the type of the stmt argument in the interceptor `Stmt*` methods still would need to be changed to not run into the issue that is described in https://github.com/ngrok/sqlmw/pull/16, right?

tcolgate commented:
@inconshreveable @fho 
I've pushed a first stab at the tests. They aren't great but...
- Extended the exists tests to stress RowsNext more
- Added a direct test of wrapRows, testing internal functions isn't generally great, but this is the easiest way to ensure the interfaces match up.
- Added a more complete test of a pgx-like fake driver that ensure the rows description stuff gets called.

I wont be able to act on feedback until Wednesday, but hopefully we are close.

inconshreveable reviewed (approved):
LGTM. will merge pending @fho's approval

fho commented on conn.go:
```diff
@@ -94,7 +94,7 @@ func (c wrappedConn) Query(query string, args []driver.Value) (driver.Rows, erro
 			return nil, err
 		}
-		return wrappedRows{intr: c.intr, parent: rows}, nil
+		return wrapRows(nil, c.intr, rows), nil //nolint
```
If I understand it right, the nil context that is passed here is then forwarded to the intercepted rows methods?
If this is the case this will cause nil ptr dereference panics for users.
Users of the lib can assume that ctx is never nil because the stdlib context document says it must never be nil. (https://pkg.go.dev/context)
```suggestion
		return wrapRows(context.Background(), c.intr, rows), nil //nolint
```

fho commented on rows.go:
```diff
@@ -3,19 +3,34 @@ package sqlmw
-	_ driver.RowsColumnTypeNullable         // TODO
-	_ driver.RowsColumnTypePrecisionScale   // TODO
-	_ driver.RowsColumnTypeScanType         // TODO
-	_ driver.RowsNextResultSet              // TODO
+	_ driver.Rows = wrappedRows{}
```
optional:
To make it consistent, we should add similiar assignments for all newly added wrappedRows types, alternatively we could also remove the last remaining `_ driver.Rows = wrappedRows{}` assignment.

fho commented on tools/rows_picker_gen.go:
```diff
@@ -0,0 +1,145 @@
+			log.Fatalf("could not create file %q, %v", *fn, err)
+		}
+	}
+	defer out.Close()
```
The return value of Close() should be checked, if it is not nil the program should exit with an error.
In the unlikely case that Close() failed the content might have not be written to disk.

fho commented on rows_test.go:
```diff
@@ -75,3 +77,285 @@ func TestRowsClose(t *testing.T) {
+			_, rok := rows.(driver.RowsNextResultSet)
+			_, wok := wr.(driver.RowsNextResultSet)
+			if rok != wok {
+				t.Fatalf("inconsinstent support for driver.RowsNextResultSet")
```
optional:

```suggestion
				t.Fatalf("inconsistent support for driver.RowsNextResultSet")
```
That typo occurs multiple times. :-)

fho commented on rows_test.go:
```diff
@@ -75,3 +77,285 @@ func TestRowsClose(t *testing.T) {
+		rows: rows,
+	}
+	con.stmt = stmt
+	driverName := t.Name()
```
optional:

If `go test` is called with a `-count` parameter with a value >1 the testcases will fail because a driver with the same name is registered multiple times.
The test name is not unique enough, adding a microsec timestamp or a uuid could fix it.

The issue is not introduced with this PR and exists also in other testcases.
Therefore I don't see it as strictly necessary to be changed for the new testcases in this PR.

tcolgate replied:
One approach I've used for this kind of thing before is to have an atomic somewhere that you increment when generating a driver name.

fho replied:
that approach is much better then my suggestion :-)

fho commented on rows_test.go:
```diff
@@ -75,3 +77,285 @@ func TestRowsClose(t *testing.T) {
+	for rs.Next() {
+		err := rs.Scan(&id, &name)
+		if err != nil {
+			log.Fatal(err)
```
```suggestion
			t.Fatal(err)
```

fho commented on rows_test.go:
```diff
@@ -75,3 +77,285 @@ func TestRowsClose(t *testing.T) {
+	for qrs.Next() {
+		err := qrs.Scan(&id, &name)
+		if err != nil {
+			log.Fatal(err)
```
```suggestion
			t.Fatal(err)
```

fho commented on rows_test.go:
```diff
@@ -75,3 +77,285 @@ func TestRowsClose(t *testing.T) {
+		t.Errorf("wrong column name or types count")
+	}
+
+	// There's no real way these can be called, but we'll text incase the
```
```suggestion
	// There's no real way these can be called, but we'll test in case the
```

inconshreveable commented:
ready to merge after tests pass

tcolgate commented:
I will finish the review comments on Tuesday

tcolgate commented:
Hopefully all good to go now?

inconshreveable commented:
ty @tcolgate !
